### PR TITLE
Add the ability to trigger builds with tasks

### PR DESCRIPTION
**Pull request by davemcorwin**
This PR grew a bit, i'll try to provide some guidance.

Overall, the intention was to create a generic "application pool" interface that has 2 implementations, `CFApplicationPool` for managing builder instances and `CFTaskPool` for managing tasks. This required a decent amount of refactoring in the code and tests as it moved more towards a "dependency injection" type architecture.

This is deployable alongside an existing builder as long as different sqs queues are used.

- [x] add `sqs_task_url` to the sqs queue user provided service in PRODUCTION.

This also includes:
- more env-related refactoring
- multiple manifests so it can be deployed using either application pool
- some other simplifications along the way (cf auth client)

### CODE REVIEW

File: src/cloud-foundry-api-client.js [hunk 1]
```diff
@@ -57,69 +47,85 @@ class CloudFoundryAPIClient {
...
+    return this._authRequest(
+      'PUT',
+      container.url,
+      { environment_json: environment }
+    )
+      .then(() => this._authRequest('POST', `${container.url}/restage`));
+  }
+
+  _buildContainerNames(buildContainerBaseName, numBuildContainers) {
+    if (numBuildContainers <= 1) {
```
Comment (amirbey) [inline src/cloud-foundry-api-client.js]: if numbBuildContainers is 0 should we return non-empty array?

File: app.js [hunk 2]
```diff
@@ -1,18 +1,23 @@
...
+} = process.env;
 
 // If settings present, start New Relic
 if (NEW_RELIC_APP_NAME && NEW_RELIC_LICENSE_KEY) {
   require('newrelic'); // eslint-disable-line global-require
 }
 
-const builderPool = new Cluster();
+// eslint-disable-next-line no-use-before-define
+const BuilderPool = getBuilderPool(appEnv.builderPoolType);
```
[the remark this answers is not in the document]
Comment (davemcorwin) [reply app.js]: The eslint comment or the builder pool? I will move the function definition up so the eslint comment is not necessary. The `BuilderPool` is used on the next line.

File: env.js [hunk 3]
```diff
@@ -24,6 +33,21 @@ appEnv.cloudFoundryApiHost = cfApiHost;
...
+appEnv.buildTimeout = 1000 * (parseInt(BUILD_TIMEOUT_SECONDS, 10) || 21 * 60); // milliseconds
+
+// Application Builder Pool
+appEnv.buildContainerBaseName = BUILD_CONTAINER_BASE_NAME;
+appEnv.numBuildContainers = parseInt(NUM_BUILD_CONTAINERS, 10);
+
+// Task Builder Pool
+appEnv.taskAppName = TASK_APP_NAME;
+appEnv.taskAppCommand = TASK_APP_COMMAND;
+appEnv.taskDisk = TASK_DISK_GB && TASK_DISK_GB * 1024;
```
[the remark this answers is not in the document]
Comment (davemcorwin) [reply env.js]: If not specified, `TASK_DISK_GB` will be `undefined` and `appEnv.taskDisk` will also be undefined. This is meant to be required. Maybe it's worthwhile calling `parseInt` here though.

File: manifests/production/task-manifest.yml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+---
+applications:
+- name: federalist-builder-task
+  routes:
+  - route: federalist-builder-task.fr.cloud.gov
+  disk_quota: 512M
+  memory: 128MB
```
[the remark this answers is not in the document]
Comment (davemcorwin) [reply manifests/production/task-manifest.yml]: Changing to `M`

File: manifests/staging/task-manifest.yml [hunk 5]
```diff
@@ -0,0 +1,21 @@
+---
+applications:
+- name: federalist-builder-staging-task
+  routes:
+  - route: federalist-builder-staging-task.fr.cloud.gov
+  disk_quota: 512M
+  memory: 128MB
```
[the remark this answers is not in the document]
Comment (davemcorwin) [reply manifests/staging/task-manifest.yml]: Changing to `M`

File: src/cf-task-pool.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  async startBuild(build) {
+    const buildTask = this._buildTask(build);
+    try {
+      const task = await this._apiClient.startTaskForApp(buildTask, this._taskAppGUID);
+      logger.info('Started build %s in task %s guid %s', build.buildID, task.name, task.guid);
+      this._builds[build.buildID] = {
+        taskGUID: task.guid,
+        timeout: this._createBuildTimeout(build),
+      };
+      return undefined;
```
[the remark this answers is not in the document]
Comment (davemcorwin) [reply src/cf-task-pool.js]: The `return undefined;` makes the return explicit and satisfies a linting constraint. Or so I thought, I'm not seeing the linting error right now, but either way, the contract is that this resolves to `undefined` which I think both implementations do.

File: src/cloud-foundry-api-client.js (hunk 1 above)
Comment (davemcorwin) [reply src/cloud-foundry-api-client.js]: My idea here was that we probably won't specify the number of containers unless it is more than 1 so just assume we want 1.

**Comment by davemcorwin** [comment]
@amirbey thanks for the comments, pushed up some changes per your feedback!

File: app.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (amirbey) [reply app.js]: forgive me ... i don't konw how I missed that :)

**Pull request merged**